### PR TITLE
Indigo shade related fixes and other changes

### PR DESCRIPTION
**Pull request by PepperCode1**
- Apply disabled shade from vanilla quads directly to material. `MutableQuadViewImpl#fromVanilla` now modifies the passed material to force disable diffuse if the vanilla quad has shade disabled, which allows the material to accurately represent the diffuse state of the quad. `QuadViewImpl` no longer stores a hidden `shade` boolean that cannot be copied by a FRAPI user. These changes do not change the behavior of any Indigo methods.
- Fix `AoCalculator` using AO face data computed with a potentially different shade state. With enhanced AO mode, after a non-axis-aligned quad had AO data calculated and cached, any future quads on the same block with a different shade state would be rendered as though they had the shade state of the first quad. The AO data cache now separates data by shade state to fix this issue.
- Respect cull parameter passed to non-terrain rendering.
- Other minor cleanup and performance improvements

### CODE REVIEW

File: fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/mesh/MutableQuadViewImpl.java [hunk 1]
```diff
@@ -116,9 +117,13 @@ public final MutableQuadViewImpl fromVanilla(BakedQuad quad, RenderMaterial mate
 		data[baseIndex + HEADER_BITS] = EncodingFormat.cullFace(0, cullFace);
 		nominalFace(quad.getFace());
 		colorIndex(quad.getColorIndex());
+
+		if (!quad.hasShade()) {
```
Comment (Technici4n) [inline fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/mesh/MutableQuadViewImpl.java:121]: Note to self: When the quad has shade and the material does not, an "opinionated" choice needs to be made on whether to shade or not here. I'm fine with the current behavior of disabling the shade if either the quad or the material disables it.

File: fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/BlockRenderInfo.java [hunk 2]
```diff
@@ -84,7 +94,24 @@ int blockColor(int colorIndex) {
 	}
 
 	boolean shouldDrawFace(Direction face) {
```
Comment (Technici4n) [inline fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/BlockRenderInfo.java]: Mark it as `@Nullable` maybe?

File: fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/AbstractQuadRenderer.java [hunk 3]
```diff
@@ -235,9 +234,9 @@ int flatBrightness(MutableQuadViewImpl quad, BlockState blockState, BlockPos pos
 	 * even for un-shaded quads. These are also applied with AO shading but that is done in AO calculator.
 	 */
 	private void shadeFlatQuad(MutableQuadViewImpl quad) {
-		if ((quad.geometryFlags() & GeometryHelper.AXIS_ALIGNED_FLAG) == 0 || quad.hasVertexNormals()) {
-			// Quads that aren't direction-aligned or that have vertex normals need to be shaded
-			// using interpolation - vanilla can't handle them. Generally only applies to modded models.
+		if (quad.hasVertexNormals()) {
+			// Quads that have vertex normals need to be shaded using interpolation - vanilla can't
+			// handle them. Generally only applies to modded models.
```
Comment (Technici4n) [inline fabric-renderer-indigo/src/client/java/net/fabricmc/fabric/impl/client/indigo/renderer/render/AbstractQuadRenderer.java:239]: Why?

File: fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/impl/renderer/RendererHolder.java [hunk 4]
```diff
@@ -16,17 +16,30 @@
...
 import net.fabricmc.fabric.api.renderer.v1.Renderer;
-import net.fabricmc.fabric.api.renderer.v1.RendererAccess;
 
-public final class RendererAccessImpl implements RendererAccess {
-	public static final RendererAccessImpl INSTANCE = new RendererAccessImpl();
+public final class RendererHolder {
+	@Nullable
+	private static Renderer activeRenderer = null;
+
+	/** avoids null test every call to {@link #isPresent()}. */
```
Comment (Technici4n) [inline fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/impl/renderer/RendererHolder.java]: That's weird, `null` checks should largely be fast enough. But it was already there so /shrug.

File: fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/Renderer.java [hunk 5]
```diff
@@ -23,13 +23,41 @@
...
  * for model lighting, buffering and rendering. Such plug-ins implement the
  * enhanced model rendering interfaces specified by the Fabric API.
  */
 public interface Renderer {
+	/**
+	 * Access to the current {@link Renderer} for creating and retrieving model builders
+	 * and materials. Will return null if no render plug in is active.
+	 */
+	@Nullable
+	static Renderer get() {
```
Comment (Technici4n) [inline fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/Renderer.java]: Could we maybe have `getNullable` that has the current behavior and `get` that throws an exception if there is no registered renderer?

File: fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java [hunk 6]
```diff
@@ -18,32 +18,43 @@
...
+	RendererAccess INSTANCE = new RendererAccess() { };
 
 	/**
 	 * Rendering extension mods must implement {@link Renderer} and
 	 * call this method during initialization.
 	 *
 	 * <p>Only one {@link Renderer} plug-in can be active in any game instance.
 	 * If a second mod attempts to register this method will throw an UnsupportedOperationException.
+	 *
+	 * <p>Use {@link Renderer#register(Renderer)} instead.
```
Comment (Technici4n) [inline fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java]:
```suggestion
	 * @deprecated Use {@link Renderer#register(Renderer)} instead.
```

File: fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java [hunk 7]
```diff
@@ -18,32 +18,43 @@
...
-	void registerRenderer(Renderer plugin);
+	default void registerRenderer(Renderer plugin) {
+		Renderer.register(plugin);
+	}
 
 	/**
 	 * Access to the current {@link Renderer} for creating and retrieving model builders
 	 * and materials. Will return null if no render plug in is active.
+	 *
+	 * <p>Use {@link Renderer#get()} instead.
```
Comment (Technici4n) [inline fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java]:
```suggestion
	 * @deprecated Use {@link Renderer#get()} instead.
```

File: fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java [hunk 8]
```diff
@@ -18,32 +18,43 @@
...
 	@Nullable
-	Renderer getRenderer();
+	default Renderer getRenderer() {
+		return Renderer.get();
+	}
 
 	/**
 	 * Performant test for {@link #getRenderer()} != null.
+	 *
+	 * <p>Use {@link Renderer#isPresent()} instead.
```
Comment (Technici4n) [inline fabric-renderer-api-v1/src/client/java/net/fabricmc/fabric/api/renderer/v1/RendererAccess.java]:
```suggestion
	 * @deprecated Use {@link Renderer#isPresent()} instead.
```

**Review by Technici4n — CHANGES_REQUESTED** [review]
Looks very good overall, just a few comments.

**Review by Technici4n — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]

**Comment by modmuss50** [comment]
Note to future self: Backport to 1.18.2 and 1.19.2

**Review by Player3324 — APPROVED** [review]

**Pull request merged**